src/vsr/replica_format.zig

```diff
@@ -26,195 +22,287 @@ pub fn format(
+                // Direct I/O requires the buffer to be sector-aligned. Allocate a buffer for each
+                // sector in the arena, so they can be written concurrently.
+                const header_buffer = try arena.alignedAlloc(
+                    u8,
+                    constants.sector_size,
+                    constants.sector_size,
+                );
+                @memset(header_buffer, 0);
+
+                const header, const prepare_offset = wal_header(cluster, slot, header_buffer);
+                assert(!stdx.zeroed(header_buffer[0..@sizeOf(Header.Prepare)]));
+                assert(stdx.zeroed(header_buffer[@sizeOf(Header.Prepare)..]));
+                assert(header.valid_checksum());
+
+                if (header.op == 0) {
+                    assert(header.operation == .root);
+                } else {
+                    assert(header.operation == .reserved);
                 }
 
                 storage.write_sectors(
                     write_sectors_callback,
-                    &self.write,
-                    wal_buffer[0..size],
+                    &self.writes[self.writes_pending],
+                    header_buffer,
                     .wal_prepares,
-                    wal_offset,
+                    prepare_offset,
                 );
-                self.formatting = true;
-                while (self.formatting) storage.run();
-                wal_offset += size;
+                self.writes_pending += 1;
             }
-            // There are no prepares left to write.
-            assert(format_wal_prepares(cluster, wal_offset, wal_buffer) == 0);
-
-            wal_offset = 0;
-            while (wal_offset < constants.journal_size_headers) {
-                const size = format_wal_headers(cluster, wal_offset, wal_buffer);
-                assert(size > 0);
-
-                for (std.mem.bytesAsSlice(Header.Prepare, wal_buffer[0..size])) |*header| {
-                    assert(header.valid_checksum());
-
-                    if (header.op == 0) {
-                        assert(header.operation == .root);
-                    } else {
-                        assert(header.operation == .reserved);
-                    }
-                }
 
-                storage.write_sectors(
-                    write_sectors_callback,
-                    &self.write,
-                    wal_buffer[0..size],
-                    .wal_headers,
-                    wal_offset,
+            // Direct I/O requires the buffer to be sector-aligned. Unlike the Prepares above that
+            // require a buffer per prepare, since they are spread out with zeros inbetween, the
+            // headers zone is contiguous so a single buffer will do.
+            const headers_buffer = try arena.alignedAlloc(
+                u8,
+                constants.sector_size,
+                vsr.sector_ceil(constants.journal_size_headers),
+            );
+            @memset(headers_buffer, 0);
```

Comment:
In this memset and the one below, lets be more intentional and zero out only the padding bytes. Blank `memset` has a higher chance of being correct by accident, but it is hard to reason about, as it doesn't informed the user which specific bytes are padding.